File: matchsticks-to-square.py
class Solution:
    def isPossible(self, i, sides, matchsticks, n, total):
        if i >= n:
            return len(set(sides)) == 1
        key = tuple(sorted(sides) + [i])
        if key in self.cache:
            return self.cache[key]
        for j in range(4):
            if 4 * (sides[j] + matchsticks[i]) > total:
                continue
            sides[j] += matchsticks[i]
            curr = self.isPossible(i + 1, sides, matchsticks, n, total)
            if curr:
                self.cache[key] = True
                return True
            else:
                sides[j] -= matchsticks[i]
        self.cache[key] = False
        return False
    
    def makesquare(self, matchsticks: List[int]) -> bool:
        matchsticks.sort(reverse = True)
        n = len(matchsticks)
        total = sum(matchsticks)
        self.cache = {}
        sides = [0, 0, 0, 0]
        return self.isPossible(0, sides, matchsticks, n, total)

File: matchsticks-to-square.cpp
class Solution {
    bool check(vector<int> &sides,const vector<int> &matches, int index, int target) {
        if (index == matches.size())
            return sides[0] == sides[1] && sides[0] == sides[2] && sides[0] == sides[3];
        for (int i = 0; i < 4; ++i) {
            if (sides[i] + matches[index] > target) continue;
            sides[i] += matches[index];
            if (check(sides, matches, index + 1, target))
                return true;
            sides[i] -= matches[index];
        }
        return false;
    }
public:
    bool makesquare(vector<int>& nums) {
        if (nums.empty()) return false;
        int s = 0;
        for (int el : nums) s += el;
        if (s % 4 != 0) return false;
        int target = s / 4;
        sort(nums.begin(), nums.end(), greater<>());
        vector<int> sides(4, 0);
        return check(sides, nums, 0, target);
    }
};